File: SpaceInvaders/Composite/Composite.cs
using System;
using System.Diagnostics;

namespace SpaceInvaders
{
    public abstract class Composite : GameObject
    {
        protected CompositeName compositeName;
        protected DLink pHead;
        protected int numChildren;

        protected Composite(GameObject.Name gameName)
            : base(gameName, Sprite.Name.NullSprite)
        {
            this.numChildren = 0;
            this.pHead = null;
            this.containerType = Container.COMPOSITE;
        }

        protected Composite(GameObject.Name gameName, Sprite.Name spriteName)
            : base(gameName, spriteName)
        {
            this.numChildren = 0;
            this.pHead = null;
            this.containerType = Container.COMPOSITE;
        }


        public void Set(CompositeName name)
        {
            this.compositeName = name;
        }

        override public void Add(Component pComponent)
        {
            Debug.Assert(pComponent != null);
            DLink.AddToFront(ref this.pHead, pComponent);

            this.numChildren += 1;

            pComponent.pParent = this;
        }

        override public void Remove(Component pComponent)
        {
            Debug.Assert(pComponent != null);
            DLink.RemoveFromList(ref this.pHead, pComponent);

            this.numChildren -= 1;
        }

        public override void Print()
        {
            Debug.WriteLine(" GameObject Name: {0} ({1})", this.GetName(), this.GetHashCode());
            base.Print();

            DLink pNode = pHead;
            while (pNode != null)
            {
                pNode.Print();
                pNode = pNode.GetNext();
            }
        }

        public override void Update()
        {

            if (this.GetFirstChild() != null)
            {
                base.BaseUpdateBoundingBox();
                base.Update();
            }
           
        }

        public override void Move(float xDelta, float yDelta)
        {
            DLink pNode = this.pHead;

            while (pNode != null)    
            {
                // Get component
                Component pComponent = (Component)pNode;

                // Move component
                pComponent.Move(xDelta, yDelta);

                // Increment pointer along list
                pNode = pNode.GetNext();
            }

        }

        override public Component GetFirstChild()
        {
            DLink pNode = this.pHead;

            return (Component)pNode;
        }

        public int GetNumOfChildren()
        {
            return numChildren;
        }

        public Composite GetChildByIndex(int index)
        {
            Debug.Assert(index >= 0);
            Debug.Assert(index < numChildren);

            DLink pNode = this.pHead;
            int count = 0;

            while(count <= index)
            {
                if (count == index)
                {
                    return (Composite)pNode;  
                }

                pNode = pNode.GetNext();
                count++;
            }

            return null;
        }

        public CompositeName GetCompositeName()
        {
            return this.compositeName;
        }

        public void SetName(CompositeName name)
        {
            this.compositeName = name;
        }

        public DLink GetComponentList()
        {
            return this.pHead;
        }

        public enum CompositeName
        {
            AlienGrid,
            AlienCol1,
            AlienCol2,
            AlienCol3,
            AlienCol4,
            AlienCol5,
            AlienCol6,
            AlienCol7,
            AlienCol8,
            AlienCol9,
            AlienCol10,
            AlienCol11,
            MissileGroup,
            CoreCannonGroup,
            ShieldGroup,
            Uninitialized,
            BombRoot,
            FlyingSaucerRoot
        };

    }
}

File: quarkus-migration/space-invaders-quarkus/src/main/java/net/beeland/spaceinvaders/composite/Composite.java
package net.beeland.spaceinvaders.composite;

import net.beeland.spaceinvaders.gameobject.GameObject;
import net.beeland.spaceinvaders.manager.DLink;

/**
 * Composite is an abstract class that extends GameObject and can contain children.
 * This implements the Composite pattern, allowing game objects to form hierarchies.
 * 
 * Examples include:
 * - AlienGrid containing AlienColumns
 * - AlienColumn containing individual Aliens
 * - ShieldGroup containing Shield structures
 */
public abstract class Composite extends GameObject {
    
    protected CompositeName compositeName;
    protected DLink head;
    protected int numChildren;

    /**
     * Creates a composite game object with the specified name.
     *
     * @param gameName The game object name
     */
    protected Composite(GameObjectName gameName) {
        super(gameName);
        this.numChildren = 0;
        this.head = null;
        this.containerType = ContainerType.COMPOSITE;
        this.compositeName = CompositeName.UNINITIALIZED;
    }

    /**
     * Sets the composite name for this composite.
     *
     * @param name The composite name
     */
    public void setCompositeName(CompositeName name) {
        this.compositeName = name;
    }

    /**
     * Gets the composite name.
     *
     * @return The composite name
     */
    public CompositeName getCompositeName() {
        return this.compositeName;
    }

    /**
     * Adds a child component to this composite.
     * The child is added to the front of the list.
     *
     * @param component The component to add
     * @throws IllegalArgumentException if component is null
     */
    @Override
    public void add(Component component) {
        if (component == null) {
            throw new IllegalArgumentException("Cannot add null component");
        }

        // Add to front of list
        if (this.head == null) {
            this.head = component;
            component.clear();
        } else {
            component.setNext(this.head);
            this.head.setPrev(component);
            component.setPrev(null);
            this.head = component;
        }

        this.numChildren++;
        component.setParent(this);
    }

    /**
     * Removes a child component from this composite.
     *
     * @param component The component to remove
     * @throws IllegalArgumentException if component is null
     */
    @Override
    public void remove(Component component) {
        if (component == null) {
            throw new IllegalArgumentException("Cannot remove null component");
        }

        // Remove from list
        if (component == this.head) {
            this.head = component.getNext();
            if (this.head != null) {
                this.head.setPrev(null);
            }
        } else {
            component.remove();
        }

        this.numChildren--;
    }

    /**
     * Gets the first child component.
     *
     * @return The first child, or null if no children
     */
    @Override
    public Component getFirstChild() {
        return (Component) this.head;
    }

    /**
     * Gets the number of children in this composite.
     *
     * @return The number of children
     */
    public int getNumChildren() {
        return this.numChildren;
    }

    /**
     * Gets a child by index.
     *
     * @param index The index (0-based)
     * @return The child at the specified index
     * @throws IndexOutOfBoundsException if index is invalid
     */
    public Component getChildByIndex(int index) {
        if (index < 0 || index >= numChildren) {
            throw new IndexOutOfBoundsException("Index " + index + " out of bounds for " + numChildren + " children");
        }

        DLink node = this.head;
        int count = 0;

        while (count < index && node != null) {
            node = node.getNext();
            count++;
        }

        return (Component) node;
    }

    /**
     * Gets the head of the component list.
     *
     * @return The head of the list
     */
    public DLink getComponentList() {
        return this.head;
    }

    /**
     * Moves this composite and all its children by the specified delta.
     *
     * @param xDelta The amount to move in the X direction
     * @param yDelta The amount to move in the Y direction
     */
    @Override
    public void move(float xDelta, float yDelta) {
        DLink node = this.head;

        while (node != null) {
            Component component = (Component) node;
            component.move(xDelta, yDelta);
            node = node.getNext();
        }
    }

    /**
     * Updates this composite and all its children.
     * Recalculates the bounding box based on children.
     */
    @Override
    public void update() {
        if (getFirstChild() != null) {
            baseUpdateBoundingBox();
            super.update();
        }
    }

    /**
     * Prints debug information about this composite and its children.
     */
    public void print() {
        System.out.println("Composite: " + getName() + " (" + compositeName + ") - " + numChildren + " children");
        
        DLink node = head;
        while (node != null) {
            if (node instanceof Component) {
                Component comp = (Component) node;
                if (comp instanceof GameObject) {
                    System.out.println("  Child: " + ((GameObject) comp).getName());
                }
            }
            node = node.getNext();
        }
    }

    /**
     * Enum defining all possible composite names in the game.
     */
    public enum CompositeName {
        ALIEN_GRID,
        ALIEN_COL_1,
        ALIEN_COL_2,
        ALIEN_COL_3,
        ALIEN_COL_4,
        ALIEN_COL_5,
        ALIEN_COL_6,
        ALIEN_COL_7,
        ALIEN_COL_8,
        ALIEN_COL_9,
        ALIEN_COL_10,
        ALIEN_COL_11,
        MISSILE_GROUP,
        CORE_CANNON_GROUP,
        SHIELD_GROUP,
        BOMB_ROOT,
        FLYING_SAUCER_ROOT,
        UNINITIALIZED
    }
}